### PR TITLE
Map out block validation data flow

### LINKED ISSUE #529: Map out dataflow dependencies for block verification.
To implement the block verifier (#428), we need to have mapped out the data flow dependencies of all of the components of block validation.  This will allow us to split the validation work into tasks with explicit dependencies.

### PR DESCRIPTION
**Pull request by dconnolly**
Part of #529

### CODE REVIEW

**Comment by dconnolly** [comment]
https://github.com/ZcashFoundation/zebra/issues/529#issuecomment-648996923

**Comment by dconnolly** [comment]
That build failure is very weird: 

image

**Comment by dconnolly** [comment]
> That build failure is very weird:
> 
> image

I wonder if crates.io is having a hard time with new Rust 1.45.0 release?

**Comment by dconnolly** [comment]
What the hell

image

File: design/data-flow.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
+- Started in `ProcessNewBlock`, `main.cpp:4258`
+    - Calls `CheckBlock`, defined at  https://github.com/zcash/zcash/blob/ab2b7c0969391d8a57d90d008665da02f3f618e7/src/main.cpp#L3930
+        - Calls `CheckBlockHeader` https://github.com/zcash/zcash/blob/ab2b7c0969391d8a57d90d008665da02f3f618e7/src/main.cpp#L3900
+            - checks that the block version is not too old
+                - this requires no information
+            - checks that the equihash solution is valid
+                - https://github.com/zcash/zcash/blob/ab2b7c0969391d8a57d90d008665da02f3f618e7/src/pow.cpp#L96
+                - requires no information except `n`, `k` params
+            - checks that the proof of work parameters are valid
+                - requires the current proof of work amount `params.powLimit`
```
Comment (teor2345) [inline design/data-flow.md:10]: Added to the consensus parameters tracking issue #565

File: design/data-flow.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+                            - Moar: https://github.com/zcash/zcash/blob/ab2b7c0969391d8a57d90d008665da02f3f618e7/src/main.cpp#L1091
+        - Sum up "LegacySigOps" for each transaction and check that it's less than some maximum 
+    - Acquires a lock, then calls `MarkBlockAsReceived` (networking?)
+    - Calls `AcceptBlock`, defined at: https://github.com/zcash/zcash/blob/ab2b7c0969391d8a57d90d008665da02f3f618e7/src/main.cpp#L4180
+        - Calls `AcceptBlockHeader`, defined at: https://github.com/zcash/zcash/blob/ab2b7c0969391d8a57d90d008665da02f3f618e7/src/main.cpp#L4130
+            - Also calls `CheckBlockHeader`
+            - Calls `ContextualCheckBlockHeader`, defined at: https://github.com/zcash/zcash/blob/ab2b7c0969391d8a57d90d008665da02f3f618e7/src/main.cpp#L3900
+                - Does checks given a pointer to the previous block
+                - Check Equihash solution is valid
+                    - In our code we compute the equihash solution on the block alone, we will need to also do a step to check that its block height is the appopriate N+1 re: the previous block
```
Comment (teor2345) [inline design/data-flow.md:35]: Fixed in PR #748

File: design/data-flow.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+                - Check proof of work
```
Comment (teor2345) [inline design/data-flow.md:36]: See #572 and PR #799

File: design/data-flow.md [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+                - Check timestamp against prev
+                - Check future timestamp soft fork rule introduced in v2.1.1-1.
+                    - https://github.com/zcash/zcash/blob/ab2b7c0969391d8a57d90d008665da02f3f618e7/src/main.cpp#L4027
+                - Check timestamp
+                    - https://github.com/zcash/zcash/blob/ab2b7c0969391d8a57d90d008665da02f3f618e7/src/main.cpp#L4040
+                - Reject block.nVersion < 4 blocks
```
Comment (teor2345) [inline design/data-flow.md:42]: Done as part of our parsing

File: design/data-flow.md [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+                - Don't accept any forks from the main chain prior to last checkpoint
+                    - We will probably get this 'more naturally' and don't need to explicitly check it like this
```
Comment (teor2345) [inline design/data-flow.md:44]: CheckpointVerifier effectively does this, and when we don't have hard-coded checkpoints, the parallel verification RFC should handle it #763

File: design/data-flow.md [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        - Calls `CheckBlock` also, w/ ProofVerifier::Disabled == true
+            - https://github.com/zcash/zcash/blob/ab2b7c0969391d8a57d90d008665da02f3f618e7/src/main.cpp#L4211
+        - Calls `ContextualCheckBlock`, defined at https://github.com/zcash/zcash/blob/ab2b7c0969391d8a57d90d008665da02f3f618e7/src/main.cpp#L4065
+            - For each transaction:
+                - Calls `ContextualCheckTransaction`, defined at https://github.com/zcash/zcash/blob/ab2b7c0969391d8a57d90d008665da02f3f618e7/src/main.cpp#L760
+                    - For Zebra, we should only be doing Transaction v4 (Sapling+) checks, because checkpointing
```
Comment (teor2345) [inline design/data-flow.md:50]: Do we need to keep context from v1-v3 transactions, so we can verify v4 transactions?

For example, UTXOs and revealed nullifiers.

File: design/data-flow.md [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+                    - Check Spend/Output Descriptions 
+                        - Check Spends: https://github.com/zcash/zcash/blob/ab2b7c0969391d8a57d90d008665da02f3f618e7/src/main.cpp#L998
+                        - Check Outputs: https://github.com/zcash/zcash/blob/ab2b7c0969391d8a57d90d008665da02f3f618e7/src/main.cpp#L1018
+                        - Check final balances: https://github.com/zcash/zcash/blob/ab2b7c0969391d8a57d90d008665da02f3f618e7/src/main.cpp#L1036
+                - Calls `IsFinalTx`, defined at https://github.com/zcash/zcash/blob/ab2b7c0969391d8a57d90d008665da02f3f618e7/src/main.cpp#L675
+                    - Calls tx::IsFinal(): https://github.com/zcash/zcash/blob/6d9573c66e33b5b742c49ab35e816d74d4ab55b1/src/primitives/transaction.h#L401
+            - Enforce BIP 34: https://github.com/zcash/zcash/blob/ab2b7c0969391d8a57d90d008665da02f3f618e7/src/main.cpp#L4090
+            - If blockheight < the block just before the first subsidy halving block, dispense Founder's Reward: https://github.com/zcash/zcash/blob/ab2b7c0969391d8a57d90d008665da02f3f618e7/src/main.cpp#L4104
```
Comment (teor2345) [inline design/data-flow.md:65]: The coinbase, founders reward, and dev fund don't need any verification context, so I've opened ticket #801 for them. I also added the relevant consensus parameters to #565.

File: design/data-flow.md [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+            - Everything else is state/index management after good verification
+    - Calls `CheckBlockIndex`, defined at https://github.com/zcash/zcash/blob/ab2b7c0969391d8a57d90d008665da02f3f618e7/src/main.cpp#L5125
+    - Calls `ActivateBestChain`, defined at https://github.com/zcash/zcash/blob/ab2b7c0969391d8a57d90d008665da02f3f618e7/src/main.cpp#L3513
```
Comment (teor2345) [inline design/data-flow.md:68]: See the parallel verification RFC #763

**Review by teor2345 — APPROVED** [review]
Looks great, thanks for doing this work!

I had one question about using v1-v3 transaction state to verify v4 transactions.

**Comment by yaahc** [comment]
Is this good to merge?

File: design/data-flow.md (hunk 6 above)
Comment (dconnolly) [reply design/data-flow.md:50]: So, we might have to compute the treestate including nullifier set that is correct at the final checkpoint?

image

Comment (dconnolly) [reply design/data-flow.md:50]: To answer your question, "I'm not sure"

image

**Pull request merged**